pdns/recursor_cache.cc

```diff
@@ -264,23 +263,23 @@ bool MemRecursorCache::attemptToRefreshNSTTL(const QType& qt, const vector<DNSRe
 void MemRecursorCache::replace(time_t now, const DNSName &qname, const QType& qt, const vector<DNSRecord>& content, const vector<shared_ptr<RRSIGRecordContent>>& signatures, const std::vector<std::shared_ptr<DNSRecord>>& authorityRecs, bool auth, boost::optional<Netmask> ednsmask, vState state)
 {
   d_cachecachevalid = false;
-
+  //  cerr<<"Replacing "<<qname<<" for "<< (ednsmask ? ednsmask->toString() : "everyone") << endl;
   auto key = boost::make_tuple(qname, qt.getCode(), ednsmask ? *ednsmask : Netmask());
   bool isNew = false;
   cache_t::iterator stored = d_cache.find(key);
   if (stored == d_cache.end()) {
     stored = d_cache.insert(CacheEntry(key, CacheEntry::records_t(), auth)).first;
     isNew = true;
+  }
 
-    /* don't bother building an ecsIndex if we don't have any netmask-specific entries */
-    if (ednsmask && !ednsmask->empty()) {
-      auto ecsIndexKey = boost::make_tuple(qname, qt.getCode());
-      auto ecsIndex = d_ecsIndex.find(ecsIndexKey);
-      if (ecsIndex == d_ecsIndex.end()) {
-        ecsIndex = d_ecsIndex.insert(ECSIndexEntry(qname, qt.getCode())).first;
-      }
-      ecsIndex->addMask(*ednsmask);
+  /* don't bother building an ecsIndex if we don't have any netmask-specific entries */
+  if (ednsmask && !ednsmask->empty()) {
```

Comment:
Since the only case where we might have an existing entry in `d_cache` for this exact (`qname`, `qtype`, `ednsmask`) tuple, but not one in `d_ecsIndex`, is that the entry has expired and was removed from `d_ecsIndex` but it's still waiting for garbage collection in `d_cache`, we could skip the lookup into `d_ecsIndex` and the call to `ecsIndex->addMask()` unless `isNew` or `stored->d_ttd <= now`. It would require a comment explaining why, though.